internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"maps"
+	"reflect"
+	"regexp"
+	"slices"
+	"strings"
 
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	csiv1alpha1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	csiv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/ceph/ceph-csi-operator/internal/utils"
 )
 
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=operatorconfigs,verbs=get;list;watch
+
+type DriverType string
+
+const (
+	RbdDriverType    = "rbd"
+	CephFsDriverType = "cephfs"
+	NfsDriverType    = "nfs"
+)
+
+const ownerRefAnnotationKey = "csi.ceph.io/ownerref"
+
+// A regexp used to parse driver short name and driver type from the
+// driver's full name
+var nameRegExp, _ = regexp.Compile(fmt.Sprintf(
+	`^(.*)\.(%s|%s|%s)\.csi\.ceph\.com$`,
+	RbdDriverType,
+	CephFsDriverType,
+	NfsDriverType,
+))
+
 // DriverReconciler reconciles a Driver object
 type DriverReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+// A local reconcile object tied to a single reconcile iteration
+type driverReconcile struct {
```

Comment:
To give each reconciled event its own local structure to store information on. 
This will allow us to run the entire controller concurrently without warring that each reconcile iteration will potentially overwrite the data of another reconcile iteration. 

The reconciler structure itself should only hold data that is safe to share between multiple concurrent reconcile iterations. We broke that rule on all of our other operators and that in turn prevents us from running them with a concurrency setting of more than 1

Reply:
ack.

> itself should only hold data that is safe to share between multiple concurrent reconcile iterations
- is it possible to enforce in an API contract or something, at the least maybe a comment that states this struct shouldn't use shared mem or something like that?

resolve.

Reply:
There is no issue in having shared values on `DriverReconcile` struct (I am already doing so by embedding the entirety of the `DriverReconciler` struct in it). 

The issue starts when you set non-shared values on the `DriverrReconciler` without considering the fact that it is shared between all iterations running in parallel. But that behavior is already documented in kubebuilder's docs and I see no reason to have an in-code comment. 

I hope this completes the picture.